ambry-store/src/main/java/com.github.ambry.store/BlobStore.java

```diff
@@ -167,27 +166,27 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
       synchronized (lock) {
         // Validate that log end offset was not changed. If changed, check once again for existing
         // keys in store
-        long currentIndexEndOffset = index.getCurrentEndOffset();
-        if (currentIndexEndOffset != indexEndOffsetBeforeCheck) {
+        Offset currentIndexEndOffset = index.getCurrentEndOffset();
+        if (!currentIndexEndOffset.equals(indexEndOffsetBeforeCheck)) {
           FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
           for (MessageInfo info : messageSetToWrite.getMessageSetInfo()) {
             if (index.findKey(info.getStoreKey(), fileSpan) != null) {
               throw new StoreException("Key already exists on filespan check", StoreErrorCodes.Already_Exist);
             }
           }
         }
-        // TODO (Index changes): Working under the assumption that there is only one log segment.
-        long currentLogEndOffset = log.getEndOffset().getOffset();
+        Offset currentLogEndOffset = log.getEndOffset();
         messageSetToWrite.writeTo(log);
         logger.trace("Store : {} message set written to log", dataDir);
         List<MessageInfo> messageInfo = messageSetToWrite.getMessageSetInfo();
-        ArrayList<IndexEntry> indexEntries = new ArrayList<IndexEntry>(messageInfo.size());
+        ArrayList<IndexEntry> indexEntries = new ArrayList<>(messageInfo.size());
         for (MessageInfo info : messageInfo) {
+          FileSpan fileSpan = getFileSpanForMessage(currentLogEndOffset, info.getSize());
           IndexValue value =
               new IndexValue(info.getSize(), currentLogEndOffset, (byte) 0, info.getExpirationTimeInMs());
```

Comment:
I think this is wrong. `currentLogEndOffset` is not necessarily the start of the message. This should be:
```java
new IndexValue(info.getSize(), fileSpan.getStartOffset(), (byte) 0, info.getExpirationTimeInMs());
```